### ISSUE TITLE
Provide a way to programmatically change the local `ResourceEnvironment` (`*Qualifier`)

### ISSUE THREAD

**Issue by AzimMuradov** [JetBrains/compose-multiplatform#4197]
I need to have an ability to programmatically switch between sets of resources (based on their qualifiers). I need this feature to provide my users a way for changing the app language/interface size/theme/etc.

Fortunately, it should be quite easy to implement. The only thing needed is to make [`LocalComposeEnvironment`](https://github.com/JetBrains/compose-multiplatform/blob/master/components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/ResourceEnvironment.kt#L42) `public` and maybe add some utility functions (`WithLanguage`/`WithRegion`/...).

Basically, I want to be able to programmatically and locally change `ResourceEnvironment`:

```kotlin
internal data class ResourceEnvironment(
    val language: LanguageQualifier,
    val region: RegionQualifier,
    val theme: ThemeQualifier,
    val density: DensityQualifier
)
```

If it's ok as a feature, I'm more than happy to help!

**Comment by terrakok** [comment]
You can override the density via the CompositionLocalProvider
```
CompositionLocalProvider(
    LocalDensity provides Density(1.5f)
) {
    //compose
}
```

for the other environment values the Compose doesn't provide providers yet.
this should be implemented in the Compose and the library will use it

**Comment by ryanmitchener** [comment]
I just wanted to voice my support for this enhancement. This is particularly helpful for allowing users to select their language in an application. I'm not sure if Compose will ever implement support for this because Android has alternative methods of changing the language from inside the application via configurations and/or per-app language preferences. Those existing methods do not work for iOS or Desktop and configuring the application language in-app is more prevalent on Desktop vs Android/iOS.

At the very least it would be helpful to add a function that allows passing in a language qualifier or something similar. That would allow 3rd party libraries to wrap the resources library and work around this limitation.

```kotlin
fun stringResource(resource: StringResource, language: LanguageQualifier)
```

**Comment by juhaodong** [comment]
Very Important for us!

**Comment by aniluMango** [comment]
We have an existing ios app. which is created with custom localization handling. that not fitted in compose resources

**Comment by N7ghtm4r3** [comment]
Hi, I tried to use `Locale.setDefault(Locale.forLanguageTag("tag"))` and on `Desktop` works correctly, obviuosly using the  
`composeResources` api with the specific language tag for values folder (values-en l, etc...)
I don't know if this can work on mobile platforms or on web also

**Comment by ShreyashKore** [comment]
This would be great for the web too. Currently, the only way for users to change the language of Compose website is by changing their system's language.

**Comment by afTrolle** [comment]
Hi, I'm also looking to be able to change resources programmatically or have multiple instances with different configurations, similar to @ryanmitchener.

I'm developing a fullstack solution in Kotlin and want to use the same i18n solution (StringResource handling) for both client and server.

However, I'm currently blocked because the ResourceEnvironment constructor is internal. Would it be possible to make it public? I can't use Locale.setDefault() as I'm handling multiple requests simultaneously, which would cause a race condition.

**Comment by ahmedhosnypro** [comment]
Currently, the `stringResource` function relies on `Locale.current` to determine the appropriate string translation. However, `Locale.current` always returns the default system locale and doesn't allow for dynamic changes within the app. This makes it impossible to switch languages without restarting the application.

Here's how the current implementation works:

1.  **`stringResource` calls `rememberResourceState`:**

    ```kotlin
    @Composable
    fun stringResource(resource: StringResource): String {
        // ...
        val str by rememberResourceState(resource, { "" }) { env ->
            loadString(resource, resourceReader, env)
        }
        // ...
    }
    ```

2.  **`rememberResourceState` depends on `LocalComposeEnvironment`:**

    ```kotlin
    @Composable
    internal actual fun <T> rememberResourceState(
        // ...
    ): State<T> {
        val environment = LocalComposeEnvironment.current.rememberEnvironment()
        // ...
    }
    ```

3.  **`LocalComposeEnvironment` delegates to `DefaultComposeEnvironment`, which uses `Locale.current`:**

    ```kotlin
    internal val DefaultComposeEnvironment = object : ComposeEnvironment {
        @Composable
        override fun rememberEnvironment(): ResourceEnvironment {
            val composeLocale = Locale.current // <-- Problem: Always system locale
            // ...
        }
    }
    ```

**Proposed Modification:**

To allow dynamic locale changes, I propose the following changes to `ComposeEnvironment` and `DefaultComposeEnvironment`:

```diff
--- a/ComposeEnvironment.kt
+++ b/ComposeEnvironment.kt
@@ -1,5 +1,9 @@
 internal interface ComposeEnvironment {
     @Composable
     fun rememberEnvironment(): ResourceEnvironment
+
+    fun setLocale(
+        locale: Locale,
+    )
 }
 
 val DefaultComposeEnvironment = object : ComposeEnvironment {
@@ -26,6 +30,16 @@
         //cache ResourceEnvironment unless compose environment is changed
         return environment.value
     }
+
+    // function to change the environment
+    override fun setLocale(
+        locale: Locale,
+    ) {
+        environment.value = ResourceEnvironment(
+            LanguageQualifier(locale.language),
+            RegionQualifier(locale.region),
+            environment.value.theme,
+            environment.value.density
+        )
+    }
 }
 ```

This modification introduces a `setLocale` function to `ComposeEnvironment`, allowing us to directly update the locale used by `rememberEnvironment`. Consequently, this will trigger recomposition and update the strings loaded by `stringResource` with the new locale. 



I tested  it using the reources demo (Dekstop for now)

```kotlin
@Composable
fun LocaleController() {
    var localeTextField by remember { mutableStateOf("ar") }
    val locale = LocalComposeEnvironment.current

    OutlinedTextField(
        modifier = Modifier.padding(16.dp).fillMaxWidth(),
        value = localeTextField,
        onValueChange = { localeTextField = it },
        label = { Text("Locale") },
        enabled = true,
        colors = TextFieldDefaults.colors(
            disabledTextColor = MaterialTheme.colorScheme.onSurface,
            disabledContainerColor = MaterialTheme.colorScheme.surface,
            disabledLabelColor = MaterialTheme.colorScheme.onSurface,
        )
    )

    OutlinedButton(
        onClick = {
            locale.setLocale(Locale(localeTextField))
        },
        modifier = Modifier.padding(16.dp)
    ) {
        Text("Change Locale")
    }
}
```
[Screencast from 2024-06-24 07-04-29.webm](https://github.com/JetBrains/compose-multiplatform/assets/24454782/141a068d-8413-459a-9964-e334eeb34aea)

**Comment by jershell** [comment]
is there any news?
